talk/expert/coroutines.tex

```diff
@@ -0,0 +1,413 @@
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{\texttt{co\_yield} - generator behavior}
+  \begin{block}{Generator behavior - more boiler plate code}
+    {\scriptsize
+      \begin{cppcode*}{gobble=4}
+        template<typename T>
+        struct Task {
+          struct iterator {
+            std::coroutine_handle<promise_type> handle;
+            iterator &operator++() {
+              handle.resume();
+              if (handle.done()) { handle = nullptr; }
+              return *this;
+            }
+            T const& operator*() const { return handle.promise().value_; }
+            ...
+          };
+          iterator begin() { resume(); return h_; }
+          iterator end() { return {}; }
+          ... // plus error handling is missing
+        };
+      \end{cppcode*}
+    }
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{\texttt{co\_yield} - final usage}
+  \begin{exampleblock}{Finally able to use generator nicely - \href{https://godbolt.org/z/6bbrYrss5}{\color{blue!50}godbolt}}
+    {\scriptsize
+      \begin{cppcode*}{gobble=4}
+        Task range(int first, int last) {
+          while (first != last) {
+            co_yield first++;
+          }
+        }
+        for (int i : range(0, 10)) {
+          std::cout << i << "";
+        } // 1 2 3 4 5 6 7 8 9
+      \end{cppcode*}
+    }
+  \end{exampleblock}
+  \pause
+  \begin{alertblock}{Wait a minute : 0 is missing in the output!}
+    \begin{itemize}
+    \item resume called in \mintinline{cpp}{operator++} before 0 is printed
+    \item we should pause in \mintinline{cpp}{initial_suspend}
+    \end{itemize}
+     {\scriptsize
+      \begin{cppcode*}{gobble=4,firstnumber=10}
+        struct Task {
+          struct promise_type {
+            ...
+            std::suspend_always initial_suspend() { return {}; }
+            ...
+          }
+        };
+      \end{cppcode*}
+    }
+  \end{alertblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Lazyness allows for inifinite ranges}
+  \begin{block}{Generators do not need to be finite}
+    \begin{cppcode*}{gobble=2}
+      Task range(int first) {
+        while (true) {
+          co_yield first++;
+        }
+      }
+      for (int i : range(0)) {
+        std::cout << i << "\n";
+        if (i == 10) break;
+      }
+    \end{cppcode*}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Summary on coroutine}
+  \begin{exampleblock}{Coroutines are usable in cpp20}
+    \begin{itemize}
+    \item allow to have nice generators
+    \item also very useful for async/IO code
+      \begin{itemize}
+      \item another big usecase for this we did not touch
+      \end{itemize}
+    \end{itemize}
+  \end{exampleblock}
+  \begin{alertblock}{But library support is poor for now}
+    \begin{itemize}
+    \item standard Task/Generator classes are ont provided
+    \item leading to lot of boiler plate code for the end user
+    \item may come in \cpp23
```

Comment:
There are good coroutine libraries available, such as [`CppCoro`](https://github.com/lewissbaker/cppcoro). You may want to provide a link here.